app/controllers/staff/activity_forms_controller.rb

```diff
@@ -31,14 +31,42 @@ def update
     @activity = Activity.find(params[:activity_id])
     authorize @activity
 
-    @activity.assign_attributes(activity_params)
+    # Update activity dates, after checking they are valid via format_date
+    update_activity_dates
+    # Update all attributes except dates
+    @activity.assign_attributes(activity_params.select { |param| param.match(/^((?!date).)*$/) })
+
     update_wizard_status
 
     render_wizard @activity
   end
 
   private
 
+  def update_activity_dates
+    return if activity_params.select { |param| param.match(/(.*)date(.*)/) }.empty?
+    @activity.update(planned_start_date: format_date(planned_start_date_params))
+    @activity.update(planned_end_date: format_date(planned_end_date_params))
+    @activity.update(actual_start_date: format_date(actual_start_date_params))
+    @activity.update(actual_end_date: format_date(actual_end_date_params))
+  end
+
+  def planned_start_date_params
+    {day: activity_params["planned_start_date(3i)"], month: activity_params["planned_start_date(2i)"], year: activity_params["planned_start_date(1i)"]}
+  end
```

Comment:
This could be more terse by making a `date_params(param_name)` method that does this in the general case using string concatenation, then use the regular expression above to find all of the matching parameter names and do this as an iterator. That way you also catch future date fields.